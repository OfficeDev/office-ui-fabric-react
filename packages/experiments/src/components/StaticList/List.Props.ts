import { IObjectWithKey } from 'office-ui-fabric-react/lib/Selection';
// tslint:disable-next-line:no-unused-variable
export interface IGenericList<TItem extends IObjectWithKey> {
}

export interface IGenericListProps<TItem extends IObjectWithKey> {
  className?: string;

  items: TItem[];

  itemHeight: number;

  onRenderItem: (item: TItem, index: number) => JSX.Element | null;
}