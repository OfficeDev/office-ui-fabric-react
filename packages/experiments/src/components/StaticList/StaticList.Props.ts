import { IObjectWithKey } from 'office-ui-fabric-react/lib/Selection';
import { IGenericListProps } from './List.Props';

export interface IStaticListProps<TItem extends IObjectWithKey> extends IGenericListProps<TItem> {
  listTagName?: string;
}
