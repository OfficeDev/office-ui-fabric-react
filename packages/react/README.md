# @uifabric/react

**React components for [Office UI Fabric React](https://dev.microsoft.com/fabric)**

These are not production-ready components and **should never be used in product**. This space is useful for testing new components whose APIs might change before final release.

To import React components:

```js
import { ComponentName } from '@uifabric/react';
```

Once the React component graduates to a production release, the component will be available at:

```js
import { ComponentName } from 'office-ui-fabric-react';
```
