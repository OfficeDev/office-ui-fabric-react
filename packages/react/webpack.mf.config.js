const path = require('path');
const fs = require('fs');
const webpack = require('webpack');
const getResolveAlias = require('@fluentui/scripts/webpack/getResolveAlias');
const { createServeConfig } = require('@fluentui/scripts/webpack/webpack-resources');
const { webpackMerge } = require('just-scripts');

const shared = {
  react: { singleton: true },
  'react-dom': { singleton: true },
};

const rootComponents = fs
  .readdirSync(path.join(__dirname, 'src'))
  .filter(filename => filename[0].toUpperCase() === filename[0])
  .map(filename => filename.replace(/\.tsx?/, ''));
const rootComponentsExposes = {};
for (const component of rootComponents) {
  rootComponentsExposes[`./lib/${component}`] = `@fluentui/react/src/${component}`;
}

// TODO: fork-ts-checker-webpack-plugin needs to be taught about the resolve alias -> pathmapping in order for the type checking to work
process.env.SKIP_TYPECHECK = true;

const myConfig = {
  output: {},
  entry: './empty-entrypoint.js',
  mode: 'development',
  devtool: 'cheap-module-source-map',
  optimization: {
    concatenateModules: false,
  },
  module: {
    rules: [],
  },
  plugins: [
    new webpack.container.ModuleFederationPlugin({
      name: 'fluentuiReact',
      filename: 'remoteEntry.js',
      exposes: {
        ...rootComponentsExposes,
        './lib/compat/Button': '@fluentui/react/src/compat/Button',
        '.': '@fluentui/react/src/index',
        './lib/index': '@fluentui/react/src/index',
      },
      shared,
    }),
  ],
  resolve: {
    alias: getResolveAlias(),
  },
  devServer: {
    port: 2345,
    static: './dist',
  },
};

const metaConfig = webpackMerge.merge(createServeConfig(), myConfig);

module.exports = metaConfig;
