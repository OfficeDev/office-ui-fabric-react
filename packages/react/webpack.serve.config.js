const path = require('path');
const resources = require('@uifabric/build/webpack/webpack-resources');

module.exports = resources.createServeConfig({
  entry: './src/demo/index.tsx',

  output: {
    filename: 'demo-app.js',
  },

  externals: {
    react: 'React',
    'react-dom': 'ReactDOM',
  },

  resolve: {
    alias: {
      '@fluentui/react/src': path.join(__dirname, 'src'),
      '@fluentui/react/lib': path.join(__dirname, 'lib'),
      '@fluentui/react': path.join(__dirname, 'lib'),
      'Props.ts.js': 'Props',
      'Example.tsx.js': 'Example',
    },
  },
});
