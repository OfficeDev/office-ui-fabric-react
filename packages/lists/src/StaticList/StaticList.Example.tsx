import * as React from 'react';
import { StaticList } from './StaticList';

class StaticListExample extends React.PureComponent<{}, { items: number[] }> {
  constructor(props: {}) {
    super(props);

    this.state = {
      items: new Array(25).fill(0)
    };
  }

  public render(): JSX.Element {
    const { items } = this.state;

    return <StaticList items={items}>{(_item: number, index: number) => <li key={index}>{`Item #${index}`}</li>}</StaticList>;
  }
}

export { StaticListExample };
