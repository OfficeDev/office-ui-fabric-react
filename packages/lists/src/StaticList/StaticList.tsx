import * as React from 'react';

interface IStaticListProps<T> {
  as?: keyof JSX.IntrinsicElements;
  items?: T[];
  children?: (item: T, index: number) => React.ReactNode;
}

function defaultRowRenderer<T>(item: T, index: number): JSX.Element {
  return <li key={index}>{item}</li>;
}

class StaticList<T> extends React.PureComponent<IStaticListProps<T>> {
  public render = () => {
    const { items = [], as: RootTag = 'ul', children = defaultRowRenderer } = this.props;

    return <RootTag>{items.map(children)}</RootTag>;
  };
}

export { StaticList, IStaticListProps };
