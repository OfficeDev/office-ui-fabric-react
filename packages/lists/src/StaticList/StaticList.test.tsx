import * as React from 'react';
import { StaticList } from './StaticList';
import { shallow } from 'enzyme';
import * as renderer from 'react-test-renderer';

const COUNT = 50;
let items: number[] = [];

describe('StaticList', () => {
  beforeEach(() => {
    items = new Array(COUNT).fill(0);
  });

  it('renders empty list correctly', () => {
    const mockRenderFunction = jest.fn();
    const component = renderer.create(<StaticList>{mockRenderFunction}</StaticList>);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders no children if no items provided', () => {
    const mockRenderFunction = jest.fn();
    const wrapper = shallow(<StaticList>{mockRenderFunction}</StaticList>);

    expect(mockRenderFunction).not.toHaveBeenCalled();
    expect(wrapper.children().length).toBe(0);
  });

  it('invokes render function per item provided', () => {
    const mockRenderFunction = jest.fn();

    shallow(<StaticList items={items}>{mockRenderFunction}</StaticList>);

    expect(mockRenderFunction).toHaveBeenCalledTimes(COUNT);
  });

  it('renders custom root tag if provided', () => {
    const mockRenderFunction = jest.fn();

    const wrapper = shallow(
      <StaticList as={'ol'} items={items}>
        {mockRenderFunction}
      </StaticList>
    );

    expect(wrapper.matchesElement(<ol />)).toBe(true);
  });

  it('renders child render function per item provided', () => {
    const wrapper = shallow(
      <StaticList items={items}>{(_item: number, index: number) => <li key={index}>{`Item #${index}`}</li>}</StaticList>
    );

    expect(wrapper.children().length).toBe(COUNT);
  });

  it('re-renders when items array mutates', () => {
    const wrapper = shallow(
      <StaticList items={items}>{(_item: number, index: number) => <li key={index}>{`Item #${index}`}</li>}</StaticList>
    );

    wrapper.setProps<'items'>({ items: [...items, ...items] });

    expect(wrapper.children().length).toBe(100);
  });
});
