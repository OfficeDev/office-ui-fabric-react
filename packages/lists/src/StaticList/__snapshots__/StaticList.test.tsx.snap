// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaticList renders empty list correctly 1`] = `<ul />`;
